Raise warning when calling collectives on non-member group objects

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67639

Due to BC considerations, we cannot directly error out, as that
might break existing applications. Raise warnings first to improve
debuggability.

Closes #67600

Differential Revision: [D32075151](https://our.internmc.facebook.com/intern/diff/D32075151)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

mrshenli committed:
Raise warning when calling collectives on non-member group objects

Due to BC considerations, we cannot directly error out, as that
might break existing applications. Raise warnings first to improve
debuggability.

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32075151).

mrshenli committed:
Update on "Raise warning when calling collectives on non-member group objects"


Due to BC considerations, we cannot directly error out, as that
might break existing applications. Raise warnings first to improve
debuggability.

Closes #67600

Differential Revision: [D32075151](https://our.internmc.facebook.com/intern/diff/D32075151)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32075151).

mrshenli committed:
Update on "Raise warning when calling collectives on non-member group objects"


Due to BC considerations, we cannot directly error out, as that
might break existing applications. Raise warnings first to improve
debuggability.

Closes #67600

Differential Revision: [D32075151](https://our.internmc.facebook.com/intern/diff/D32075151)

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32075151).

rohan-varma commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2626,6 +2626,11 @@ def test_nccl_barrier_device_ids_function_argument(self):
 
+    @requires_nccl()
+    @skip_if_lt_x_gpu(2)
+    def test_nccl_warn_not_in_group(self):
```
should we also test it works fine with TORCH_DISTRIBUTED_DEBUG? It should because this is in the python level, but it would be good to make sure the wrapped process group raises this error as expected. 

You can use `with_dist_debug` decorator to do it - https://github.com/pytorch/pytorch/blob/master/test/distributed/test_c10d_nccl.py#L1068

mrshenli replied:
Yep, let me add a test.

rohan-varma commented on torch/distributed/distributed_c10d.py:
```diff
@@ -271,6 +271,14 @@ def _rank_not_in_group(group: ProcessGroup):
 
 
+def _warn_not_in_group(op_name):
+    global_rank = -1 if GroupMember.WORLD is None else GroupMember.WORLD.rank()
```
GroupMember.WORLD = None should never happen right, since we always expect to have initialized a default PG? Should we throw in that case?

mrshenli replied:
I assume the following case is allowed without calling `init_process_group`?

```python
pg = ProcessGroupGloo(...)
dist.all_reduce(tensor, group=pg)
```

rohan-varma reviewed (approved):
Lgtm, I guess there are some use cases that call directly into python process group API (i.e. bypass distributed_c10d) but probably nothing we can do about that. Also feel free to take a look at some suggestions above

mrshenli committed:
Update on "Raise warning when calling collectives on non-member group objects"


Due to BC considerations, we cannot directly error out, as that
might break existing applications. Raise warnings first to improve
debuggability.

Closes #67600

Differential Revision: [D32075151](https://our.internmc.facebook.com/intern/diff/D32075151)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

mrshenli commented:
@mrshenli has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32075151).

mrshenli reviewed (commented):
Test errors on BC, Windows, XLA are irrelevant. Landing
